### PR TITLE
8308387: CLD created and unloading list sharing _next node pointer leads to concurrent YC missing CLD roots

### PR DESCRIPTION
**Pull request by xmas92**
[JDK-8307106](https://bugs.openjdk.org/browse/JDK-8307106) introduced the ability to walk the created CLDs list in the CLDG without a lock. This was primarily introduced to allow lockless concurrent CLD roots scanning for young collections in generational ZGC. However because the CLD _next node pointer is shared between the two list this can lead to a concurrent iteration of the created CLDs missing list entries. 

This change introduces a second _unloading_next node pointer which is used for the unloading CLDs list. The set_next is now maintains the invariant that it only ever unlinks is_unloading() CLDs and maintains a consistent view of the tail list for anyone reading the list concurrently.  

Testing: tier1-3 and tier1-7 with Generational ZGC

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308387](https://bugs.openjdk.org/browse/JDK-8308387): CLD created and unloading list sharing _next node pointer leads to concurrent YC missing CLD roots

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [320be537](https://git.openjdk.org/jdk/pull/14241/files/320be5372ad4071c601ac3f1ab2ff510944df5df)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14241/head:pull/14241` \
`$ git checkout pull/14241`

Update a local copy of the PR: \
`$ git checkout pull/14241` \
`$ git pull https://git.openjdk.org/jdk.git pull/14241/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14241`

View PR using the GUI difftool: \
`$ git pr show -t 14241`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14241.diff">https://git.openjdk.org/jdk/pull/14241.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14241#issuecomment-1569744670)

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderData.hpp [hunk 1]
```diff
@@ -153,15 +153,32 @@ class ClassLoaderData : public CHeapObj<mtClass> {
...
+  // CLDs. To enable the lock free concurrent iteration of the created
+  // CLDs list used by the GC for root scanning, it is important that unlinking
+  // CLDs from the created list upholds the invariant that all unlinked CLDs are
+  // is_unloading().
+  // Any insertions to the created CLDs list is done at the head and keeps
+  // the list tail invariant.
+  // CLDs are unlinked in ClassLoaderDataGraph::do_unloading() and released in
+  // ClassLoaderDataGraph::purge(), so some syncronization is required between
+  // the two to ensure that no unlinked CLD remains in the system leading
+  // to a use after free.
```
Comment (stefank) [inline src/hotspot/share/classfile/classLoaderData.hpp:167]: I talked to Axel about the wording around what the two list contained, and how there's a slight risk that someone who's not up-to-speed in this area might misinterpret parts of the comment.

What do you think about something along the lines of this suggestion?:

```suggestion
  //
  // The ClassLoaderDataGraph maintains two lists to keep track of CLDs.
  //
  // The first list [_head, _next] is where new CLDs are registered. The CLDs
  // are only inserted at the _head, and the _next pointers are never rewritten.
  // This allows GCs to concurrently walk the list while the CLDs are being
  // concurrently unlinked.
  //
  // The second list [_unloading_head, _unloading_next] is where dead CLDs get
  // moved to during class unloading. See: ClassLoaderDataGraph::do_unloading().
  // This list is never modified while other threads are iterating over it. 
  //
  // After all dead CLDs have been moved to the unloading list, there's a
  // synchronziation point (handshake) to ensure that all threads reading these
  // CLDs finish their work. This ensures that we don't have a user-after-free
  // when we later delete the CLDs. 
  //
  // And finally, when no threads are using the unloading CLDs anymore, we
  // remove them from the class unloading list and delete them. See:
  // ClassLoaderDataGraph::purge();
```

**Review by stefank — APPROVED** [review]
I think the patch looks good. I've added a suggestion for a rewording of the comment about the two list. Feel free to skip it if you liked your comment better.

File: src/hotspot/share/classfile/classLoaderData.hpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/classfile/classLoaderData.hpp]: Your comment seems more structured, I will use it. Will only update the `and the _next pointers are never rewritten.` part to reflect the fact that `unlink_next()` will update it by removing one is_unloading CLD.

**Review by stefank — APPROVED** [review]

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 2]
```diff
@@ -501,37 +494,34 @@ bool ClassLoaderDataGraph::is_valid(ClassLoaderData* loader_data) {
...
-  ClassLoaderData* data = _head;
   ClassLoaderData* prev = nullptr;
   bool seen_dead_loader = false;
   uint loaders_processed = 0;
   uint loaders_removed = 0;
 
-  data = _head;
-  while (data != nullptr) {
+  for (ClassLoaderData* data = _head; data != nullptr; data = data->next()) {
     if (data->is_alive()) {
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:503]: There used to be more code for the case where data->is_alive() so this code had this continue, but you could simplify this by having an 'else' clause and remove the continue. That would clean this up a little.

**Review by coleenp — APPROVED** [review]
Looks good.  Is there a way to add a test?

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp [hunk 3]
```diff
@@ -268,14 +269,6 @@ inline void assert_is_safepoint_or_gc() {
          "Must be called by safepoint or GC");
 }
 
-void ClassLoaderDataGraph::cld_unloading_do(CLDClosure* cl) {
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderDataGraph.cpp:271]: Thanks for noticing that we're not using this anymore and cleaning it up.  JFR used to use this.

File: src/hotspot/share/classfile/classLoaderData.hpp [hunk 4]
```diff
@@ -153,15 +153,41 @@ class ClassLoaderData : public CHeapObj<mtClass> {
...
+  // _next->_next. This allows GCs to concurrently walk the list while the CLDs
+  // are being concurrently unlinked.
+  //
+  // The second list [_unloading_head, _unloading_next] is where dead CLDs get
+  // moved to during class unloading. See: ClassLoaderDataGraph::do_unloading().
+  // This list is never modified while other threads are iterating over it.
+  //
+  // After all dead CLDs have been moved to the unloading list, there's a
+  // synchronization point (handshake) to ensure that all threads reading these
+  // CLDs finish their work. This ensures that we don't have a user-after-free
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classLoaderData.hpp:171]: use-after-free ?

**Review by dholmes-ora — APPROVED** [review]
This logic is much clearer. Thanks for fixing the bug and cleaning it up.

**Comment by dholmes-ora** [comment]
Is the YC in the synopsis meant to be GC? If so please fix in JBS and the PR. Thanks

File: src/hotspot/share/classfile/classLoaderDataGraph.cpp (hunk 2 above)
Comment (xmas92) [reply src/hotspot/share/classfile/classLoaderDataGraph.cpp:503]: Fixed

File: src/hotspot/share/classfile/classLoaderData.hpp (hunk 4 above)
Comment (xmas92) [reply src/hotspot/share/classfile/classLoaderData.hpp]: Fixed

**Comment by xmas92** [comment]
> Looks good. Is there a way to add a test?

The timing window here is really narrow. It requires the young root scan iteration to reach a dead CLD before it is unlinked by the old collection and hold on to it until after it is linked onto the unloading list, the young root scan filters out dead CLDs so this is just a couple of instructions. It is possible to make the the race more reproducible with a couple of sleeps in the VM and running frequent concurrent young collections and class unloading.

There might be a possibility to abstract the linked list implementation and make it mockable in gtests, and then have gtests which stresses concurrent reads and unlinks and asserts invariants. However that would require a rewrite that does not feel worth it.

> Is the YC in the synopsis meant to be GC? If so please fix in JBS and the PR. Thanks

YC is suppose to abbreviate Young Collection in a generational collector. But maybe YC/OC are not common enough terms to be used in a general bug report. 
I will at least extend the description on JBS to reflect the bug, not just the symptom (the crash).

**Review by stefank — APPROVED** [review]

**Review by coleenp — APPROVED** [review]

**Comment by coleenp** [comment]
> Looks good. Is there a way to add a test?

Better not to add a test if it's sensitive to timing.  thanks.

**Comment by xmas92** [comment]
Thanks for all the reviews.
/integrate

**Pull request closed without merging**